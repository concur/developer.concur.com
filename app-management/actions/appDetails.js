import 'es6-promise';
import fetch from 'isomorphic-fetch';

export const APP_DETAILS_REQUEST = 'APP_DETAILS_REQUEST';
export const APP_DETAILS_FAILURE = 'APP_DETAILS_FAILURE';
export const APP_DETAILS_SUCCESS = 'APP_DETAILS_SUCCESS';
export const APP_DETAILS_UPDATE_SUCCESS = 'APP_DETAILS_UPDATE_SUCCESS';

export function appDetailsRequest() {
  return { type: APP_DETAILS_REQUEST };
}

export function appDetailsFailure(message) {
  return {
    type: APP_DETAILS_FAILURE,
    message,
  };
}

export function appDetailsSuccess(app) {
  return {
    type: APP_DETAILS_SUCCESS,
    app,
  };
}

export function appDetailsUpdateSuccess() {
  return { type: APP_DETAILS_UPDATE_SUCCESS };
}

export function fetchAppDetails(id) {
  return (dispatch, getState) => {
    dispatch(appDetailsRequest());

    console.log(`TODO: Pass ${id} to API call`);
    const token = getState().auth.token;
    const options = {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
    };

    return fetch(`${process.env.DEVCENTER_API_ORCHESTRATION}`, options)
      .then(response => response.json())
      .then(app => dispatch(appDetailsSuccess(app)))
      .catch(err => dispatch(appDetailsFailure(err.message)));
  };
}

export function updateAppDetails(app) {
  return (dispatch, getState) => {
    dispatch(appDetailsRequest());

    const token = getState().auth.token;
    const options = {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
    };

    return fetch(`${process.env.DEVCENTER_API_ORCHESTRATION}`, options)
      .then(response => response.json())
      .then(() => {
        dispatch(appDetailsUpdateSuccess());
        dispatch(fetchAppDetails(app.id));
      })
      .catch(err => dispatch(appDetailsFailure(err.message)));
  };
}
