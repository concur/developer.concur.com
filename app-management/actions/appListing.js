import 'es6-promise';
import fetch from 'isomorphic-fetch';

export const APP_LISTING_REQUEST = 'APP_LISTING_REQUEST';
export const APP_LISTING_FAILURE = 'APP_LISTING_FAILURE';
export const APP_LISTING_SUCCESS = 'APP_LISTING_SUCCESS';

export function appListingRequest() {
  return { type: APP_LISTING_REQUEST };
}

export function appListingFailure(message) {
  return {
    type: APP_LISTING_FAILURE,
    message,
  };
}

export function appListingSuccess(apps) {
  return {
    type: APP_LISTING_SUCCESS,
    apps,
  };
}

export function fetchAppListing() {
  return (dispatch, getState) => {
    dispatch(appListingRequest());

    const token = getState().auth.token;
    const options = {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
    };

    return fetch(`${process.env.DEVCENTER_API_ORCHESTRATION}`, options)
      .then(response => response.json())
      .then(apps => dispatch(appListingSuccess(apps)))
      .catch(err => dispatch(appListingFailure(err.message)));
  };
}
