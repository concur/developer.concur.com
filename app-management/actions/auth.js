import 'es6-promise';
import fetch from 'isomorphic-fetch';
import { hashHistory } from 'react-router';
import { reset } from 'redux-form';
import auth from '../utils/auth';

export const AUTH_LOGIN_REQUEST = 'AUTH_LOGIN_REQUEST';
export const AUTH_LOGIN_FAILURE = 'AUTH_LOGIN_FAILURE';
export const AUTH_LOGIN_SUCCESS = 'AUTH_LOGIN_SUCCESS';
export const AUTH_LOGOUT = 'AUTH_LOGOUT';

export function loginRequest() {
  return { type: AUTH_LOGIN_REQUEST };
}

export function loginFailure(message) {
  return {
    type: AUTH_LOGIN_FAILURE,
    message,
  };
}

export function loginSuccess(token) {
  auth.setToken(token);
  return {
    type: AUTH_LOGIN_SUCCESS,
    token,
  };
}

export function logout() {
  auth.removeToken();
  return { type: AUTH_LOGOUT };
}

export function login(user) {
  return (dispatch) => {
    dispatch(loginRequest());

    const options = {
      method: 'POST',
      body: JSON.stringify(user),
      headers: {
        'Content-Type': 'application/json',
      },
    };

    return fetch(`${process.env.DEVCENTER_API_FORMS}/auth/login`, options)
      .then(response => response.json())
      .then((data) => {
        if (data && data.access_token) {
          dispatch(loginSuccess(data.access_token));
          dispatch(reset('login'));
          hashHistory.push('/');
        } else {
          dispatch(loginFailure('Invalid username and/or password'));
        }
      })
      .catch(err => dispatch(loginFailure(err.message)));
  };
}
