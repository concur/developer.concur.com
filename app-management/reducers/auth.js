import {
  AUTH_LOGIN_REQUEST, AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_SUCCESS, AUTH_LOGOUT,
} from '../actions/auth';

const defaultState = {
  token: null,
  isFetching: false,
  error: '',
};

function authReducer(state = defaultState, action) {
  switch (action.type) {
    case AUTH_LOGIN_REQUEST:
      return {
        ...state,
        isFetching: true,
        error: '',
      };
    case AUTH_LOGIN_FAILURE:
      return {
        ...state,
        isFetching: false,
        error: action.message,
      };
    case AUTH_LOGIN_SUCCESS:
      return {
        token: action.token,
        isFetching: false,
        error: '',
      };
    case AUTH_LOGOUT:
      return defaultState;
    default:
      return state;
  }
}

export default authReducer;
