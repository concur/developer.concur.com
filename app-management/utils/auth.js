/* eslint-env browser */

export const TOKEN_KEY = 'concur-devcenter-jwt';

const auth = {
  setToken: token => window.localStorage.setItem(TOKEN_KEY, token),
  getToken: () => window.localStorage.getItem(TOKEN_KEY),
  removeToken: () => window.localStorage.removeItem(TOKEN_KEY),
};

export default auth;
