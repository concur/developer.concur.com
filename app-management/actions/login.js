/* eslint-env browser */

import { hashHistory } from 'react-router';
import { login } from './auth';

export const LOGIN_HANDLE_INPUT_CHANGE = 'LOGIN_HANDLE_INPUT_CHANGE';
export const LOGIN_REQUEST = 'LOGIN_REQUEST';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';

export function handleInputChange(fieldName, newValue) {
  return {
    type: LOGIN_HANDLE_INPUT_CHANGE,
    fieldName,
    newValue,
  };
}

export function loginRequest() {
  return {
    type: LOGIN_REQUEST,
  };
}

export function loginFailure(message) {
  return {
    type: LOGIN_FAILURE,
    message,
  };
}

export function loginSuccess() {
  return {
    type: LOGIN_SUCCESS,
  };
}

export function fetchToken() {
  return function thunk(dispatch, getState) {
    dispatch(loginRequest());

    const { username, password } = getState().loginForm;
    const options = {
      method: 'POST',
      body: JSON.stringify({ username, password }),
      headers: {
        'Content-Type': 'application/json',
      },
    };

    window.fetch(`${process.env.API_SERVER}/auth/login`, options)
      .then(response => response.json())
      .then((data) => {
        if (data && data.access_token) {
          dispatch(login(data.access_token));
          dispatch(loginSuccess());
          hashHistory.push('/');
        } else {
          dispatch(loginFailure('No token provided'));
        }
      })
      .catch(err => dispatch(loginFailure(err.message)));
  };
}
