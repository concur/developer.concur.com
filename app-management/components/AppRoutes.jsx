import React, { PropTypes } from 'react';
import { Router, Route, IndexRoute, hashHistory } from 'react-router';

import PageContainer from './PageContainer';
import AppListing from './AppListing';
import AppDetailsPage from './AppDetailsPage';
import NewAppPage from './NewAppPage';
import LoginPage from './LoginPage';
import SignupPage from './SignupPage';
import LogoutPage from './LogoutPage';
import NotFound from './NotFound';

import { logout } from '../actions/auth';

class AppRoutes extends React.Component {
  constructor(props) {
    super(props);

    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.requireAuth = this.requireAuth.bind(this);
    this.showAuthForms = this.showAuthForms.bind(this);
    this.logout = this.logout.bind(this);
  }

  isAuthenticated() {
    const { store } = this.props;
    return !!store.getState().auth.token;
  }

  requireAuth(nextState, replaceState) {
    if (!this.isAuthenticated()) {
      replaceState('/login');
    }
  }

  showAuthForms(nextState, replaceState) {
    if (this.isAuthenticated()) {
      replaceState('/');
    }
  }

  logout(nextState, replaceState) {
    if (this.isAuthenticated()) {
      this.props.store.dispatch(logout());
    } else {
      replaceState('/login');
    }
  }

  render () {
    return (
      <Router history={hashHistory}>
        <Route path="/" component={PageContainer}>
          <IndexRoute component={AppListing} onEnter={this.requireAuth} />
          <Route path="new" component={NewAppPage} onEnter={this.requireAuth} />
          <Route path="details/:id" component={AppDetailsPage} onEnter={this.requireAuth} />
          <Route path="login" component={LoginPage} onEnter={this.showAuthForms} />
          <Route path="signup" component={SignupPage} onEnter={this.showAuthForms} />
          <Route path="logout" component={LogoutPage} onEnter={this.logout} />
          <Route path="*" component={NotFound} />
        </Route>
      </Router>
    );
  }
}

AppRoutes.propTypes = {
  store: PropTypes.shape({
    dispatch: PropTypes.func.isRequired,
  }),
};

export default AppRoutes;
