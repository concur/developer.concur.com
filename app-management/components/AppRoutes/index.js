import AppRoutes from './AppRoutes';

export default AppRoutes;
