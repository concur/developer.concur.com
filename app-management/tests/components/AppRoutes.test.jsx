import React from 'react';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { Router, Route, IndexRoute, hashHistory } from 'react-router';
import { shallowToJson } from 'enzyme-to-json';

import AppRoutes from '../../components/AppRoutes';
import appReducer from '../../reducers';
import { loginSuccess } from '../../actions/auth';

// local storage mocks needed for login testing
import LocalStorage from '../localStorage.mock';

const replaceState = jest.fn();
let store;
let routes;

describe('<AppRoutes />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    window.localStorage = LocalStorage();
    store = createStore(appReducer);
    routes = new AppRoutes({ store });
  });

  describe('isAuthenticated', () => {
    it('returns false when app is not authenticated', () => {
      expect(routes.isAuthenticated()).toBeFalsy();
    });

    it('returns true when app is authenticated', () => {
      store.dispatch(loginSuccess('a-sample-token'));

      expect(routes.isAuthenticated()).toBeTruthy();
    });
  });

  describe('requireAuth', () => {
    it('redirects to login page when the user is not logged in', () => {
      routes.requireAuth(null, replaceState);

      expect(replaceState).toHaveBeenCalledWith('/login');
    });

    it('allows access to a page when the user is logged in', () => {
      store.dispatch(loginSuccess('a-sample-token'));
      routes.requireAuth(null, replaceState);

      expect(replaceState).toHaveBeenCalledTimes(0);
    });
  });

  describe('showAuthForms', () => {
    it('redirects to login page when the user is not logged in', () => {
      routes.showAuthForms(null, replaceState);

      expect(replaceState).toHaveBeenCalledTimes(0);
    });

    it('shows auth forms when the user is logged in', () => {
      store.dispatch(loginSuccess('a-sample-token'));
      routes.showAuthForms(null, replaceState);

      expect(replaceState).toHaveBeenCalledWith('/');
    });
  });

  describe('logout', () => {
    it('redirects to login page when the user is not logged in', () => {
      routes.logout(null, replaceState);

      expect(replaceState).toHaveBeenCalledWith('/login');
    });

    it('logs the user out when the user is logged in', () => {
      store.dispatch(loginSuccess('a-sample-token'));
      routes.logout(null, replaceState);
      const { token } = store.getState().auth;

      expect(replaceState).toHaveBeenCalledTimes(0);
      expect(token).toBeNull();
    });
  });

  describe('render', () => {
    it('renders the router and routes', () => {
      const routes = shallow(<AppRoutes />);

      expect(shallowToJson(routes)).toMatchSnapshot();
    });
  });
});
