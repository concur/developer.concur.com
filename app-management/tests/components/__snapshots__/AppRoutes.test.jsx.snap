exports[`<AppRoutes /> render renders the router and routes 1`] = `
<Router
  history={
    Object {
      "createHref": [Function],
      "createKey": [Function],
      "createLocation": [Function],
      "createPath": [Function],
      "getCurrentLocation": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "listenBefore": [Function],
      "push": [Function],
      "replace": [Function],
      "transitionTo": [Function],
    }
  }
  render={[Function]}>
  <Route
    component={[Function]}
    path="/">
    <IndexRoute
      component={[Function]}
      onEnter={[Function]} />
    <Route
      component={[Function]}
      onEnter={[Function]}
      path="new" />
    <Route
      component={[Function]}
      onEnter={[Function]}
      path="details/:id" />
    <Route
      component={[Function]}
      onEnter={[Function]}
      path="login" />
    <Route
      component={[Function]}
      onEnter={[Function]}
      path="signup" />
    <Route
      component={[Function]}
      onEnter={[Function]}
      path="logout" />
    <Route
      component={[Function]}
      path="*" />
  </Route>
</Router>
`;
